ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -42,24 +44,56 @@ def __init__(self):
         self.base_waypoints = None
         self.waypoints_2d = None
         self.waypoint_tree = None
+	self.stopline_wp_ind = -1
 
         self.loop()
 
-    def publish_waypoints(self, idx):
-        lane = Lane()
-        lane.header = self.base_waypoints.header
-        lane.waypoints = self.base_waypoints.waypoints[idx : idx + LOOKAHEAD_WPS]
-        self.final_waypoints_pub.publish(lane)
-
     def loop(self):
         rate = rospy.Rate(50)
         while not rospy.is_shutdown():
             if self.pose and self.waypoint_tree:
                 # Get nearest waypoint
-                closest_waypoint_idx = self.get_nearest_waypoint_idx()
-                self.publish_waypoints(closest_waypoint_idx)
+                #closest_waypoint_idx = self.get_nearest_waypoint_idx()
+                self.publish_waypoints()
             rate.sleep()
 
+    def publish_waypoints(self):
+	final_lane = self.generate_lane()
+	self.final_waypoints_pub.publish(final_lane)
+
+
+    def generate_lane(self):
+	lane = Lane()
+	
+	closest_idx = self.get_nearest_waypoint_idx()
+	farthest_idx = closest_idx + LOOKAHEAD_WPS
+	base_waypoints_1 = self.base_waypoints.waypoints[closest_idx:farthest_idx]
+	
+	# If there is not a red light close to the car
+	# Get the waypoints like normal
+	# Else get the waypoints that stop at the red light
+	if self.stopline_wp_ind == -1 or (self.stopline_wp_ind >= farthest_idx):
```

Comment:
What will happen if `stopline_wp_ind` is bigger than 0 but less than `closest_idx`?
Maybe change the condition to the following to make it more robust?

```python
if self.stopline_wp_ind < closest_idx or self.stopline_wp_ind > farthest_idx:
```

Reply:
the tl_detector should never send a value that is behind the cars current position therefor
stopline_wp_ind is always greater than closest_idx